app/com/arpnetworking/rollups/ConsistencyChecker.java

```diff
@@ -0,0 +1,579 @@
+                .addData("task", task)
+                .addData("nOriginalSamples", nOriginalSamples)
+                .addData("nSamplesDropped", nSamplesDropped)
+                .addData("fractionalDataLoss", fractionalDataLoss)
+                .log();
+
+        /* TODO(spencerpearson, OBS-1176): re-trigger re-execution of bad datapoints, something like
+
+            if (discrepancyTooBig) {
+                _rollupManagerPool.tell(
+                    new RollupDefinition.Builder()
+                            .setSourceMetricName(task.getSourceMetricName())
+                            .setDestinationMetricName(task.getRollupMetricName())
+                            .setStartTime(task.getStartTime())
+                            .setPeriod(task.getPeriod())
+                            .build()
+                );
+            }
+
+         */
+    }
+
+    /* package private */ static SampleCounts parseSampleCounts(
+            final Task task,
+            final MetricsQueryResponse response
+    ) throws MalformedSampleCountResponse {
+        final Map<String, Long> countsByMetric = Maps.newHashMap();
+        if (response.getQueries().size() != 2) {
+            throw new MalformedSampleCountResponse("expected exactly 1 query, got " + response.getQueries().size(), response);
+        }
+        for (final MetricsQueryResponse.Query query : response.getQueries()) {
+            if (query.getResults().size() != 1) {
+                throw new MalformedSampleCountResponse("expected exactly 1 result, got " + query.getResults().size(), response);
+            }
+            final MetricsQueryResponse.QueryResult queryResult = query.getResults().get(0);
+            if (queryResult.getValues().size() != 1) {
+                throw new MalformedSampleCountResponse("expected exactly 1 value, got " + queryResult.getValues().size(), response);
+            }
+            final Optional<Object> value = queryResult.getValues().get(0).getValue();
+            if (!value.isPresent()) {
+                throw new MalformedSampleCountResponse("sample count has null value", response);
+            }
+            final long longValue;
+            try {
+                longValue = Long.parseLong(value.get().toString());
+            } catch (final NumberFormatException e) {
+                throw new MalformedSampleCountResponse(e, response);
+            }
+            countsByMetric.put(queryResult.getName(), longValue);
+        }
+        final Long sourceCount = countsByMetric.get(task.getSourceMetricName());
+        final Long rollupCount = countsByMetric.get(task.getRollupMetricName());
+        if (sourceCount == null || rollupCount == null) {
+            throw new MalformedSampleCountResponse(
+                    String.format("expected keys %s and %s", task.getSourceMetricName(), task.getRollupMetricName()),
+                    response
+            );
+        }
+        return new SampleCounts.Builder()
```

Comment:
This is TLB:

https://github.com/ArpNetworking/metrics-portal/blob/8a8b63dd9483ef74c0ae70f38280e4e58f25d688/app/com/arpnetworking/rollups/FailableMessage.java#L59